### ISSUE TITLE
Implement Dense Attention Layers in tfjs-layers

### ISSUE THREAD

**Issue by caisq** [tensorflow/tfjs#1958]
#### TensorFlow.js version
1.2.8

#### Describe the problem or feature request
Implement the equivalent of [tf.keras.layers.Attention](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Attention), with support for inference, training and (de)serialization. It's important for transformer / BERT related applications at the layers level.

**Comment by linonetwo** [comment]
So it is currently unable to run BERT in tfjs right?

**Comment by caisq** [comment]
@linonetwo This issue is mainly about tfjs-layers. If you have a BERT model saved as a TensorFlow SavedModel (not a Keras HDF5 model), you can still try converting it to tfjs (specifically a tfjs_graph_model) and run it in JavaScript.

**Comment by Ierezell** [comment]
Hi, does anyone have a status on this feature request ? 

I tried to implement it on my own but it doesn't seem to perform good (it works but accuracy is really low vs a classic mlp...)

I guess it's beause i'm using tf fonction instead of tf.layers ones ? (so no gradient registered or something like that ? )

<details>
<summary>Here is the code</summary>
<br>
<pre>
<code>
import * as tf from '@tensorflow/tfjs-node'

export interface TransformerLayerArgs  
{
  numHeads: number
  depth: number
}

class TransformerLayer extends tf.layers.Layer
{
  static className = "TransformerLayer";
  private numHeads: number
  private depth: number
  private queryDense: tf.layers.Layer
  private keyDense: tf.layers.Layer
  private valueDense: tf.layers.Layer
  private dense: tf.layers.Layer
  private dropout: tf.layers.Layer
  private norm1: tf.layers.Layer
  private norm2: tf.layers.Layer
  private ffDense1: tf.layers.Layer
  private ffDense2: tf.layers.Layer
  private inputDense: tf.layers.Layer

  scaledDotProductAttention(query: tf.Tensor, key: tf.Tensor, value: tf.Tensor, mask?: tf.Tensor)
  {
    return tf.tidy(() =>
    {
      // Should I use tf.layers.dot(...).apply() ? 
      const matmul_qk = tf.matMul(query, key, false, true)

      // Doesn't seem to be a div in tf.layers nor a sqrt... maybe with some multiplication hacks
      let logits = tf.div(matmul_qk, tf.sqrt(tf.cast(this.depth, 'float32')))

      if (mask) { logits = tf.add(logits, tf.mul(mask, 1e-9)) }

      const attention_weights = tf.softmax(logits, -1)

      return tf.matMul(attention_weights, value, true, false)
    })
  }

  splitHeads(inputs: tf.Tensor)
  {
    return tf.tidy(() =>
    {
      const batchSize = inputs.shape[0]
      // Should I use tf.layers.reshape(...).apply() ? 
      inputs = inputs.reshape([batchSize, -1, this.numHeads, this.depth])
      return tf.transpose(inputs, [0, 2, 1, 3])
    })
  }


  multiHeadAttention(query: tf.Tensor, key: tf.Tensor, value: tf.Tensor, mask?: tf.Tensor): tf.Tensor
  {
    return tf.tidy(() =>
    {
      const queryD = this.queryDense.apply(query) as tf.Tensor
      const keyD = this.keyDense.apply(key) as tf.Tensor
      const valueD = this.valueDense.apply(value) as tf.Tensor // Because of typescript tf.SymbolicTensor... 


      const queryS = this.splitHeads(queryD)
      const keyS = this.splitHeads(keyD)
      const valueS = this.splitHeads(valueD)

      const scaledAttention = this.scaledDotProductAttention(queryS, keyS, valueS, mask)
      const scaledAttentionT = tf.transpose(scaledAttention, [0, 2, 1, 3]) // Same here for tf.layers
      const concat_attention = tf.reshape(scaledAttentionT,
        [scaledAttentionT.shape[0], -1, this.depth])

      const attention = this.dense.apply(concat_attention) as tf.Tensor

      return attention
    })
  }

  constructor(args: TransformerLayerArgs)
  {
    super({})

    this.numHeads = args.numHeads
    this.depth = args.depth

    if (this.depth % this.numHeads != 0)
    {
      throw new Error(`Assertion error : depth(${this.depth}) % numHead(${this.numHeads}) != 0 `)
    }

    // Should all this be in build() method ? 
    this.queryDense = tf.layers.dense({ units: this.depth })
    this.inputDense = tf.layers.dense({ units: this.depth })
    this.keyDense = tf.layers.dense({ units: this.depth })
    this.valueDense = tf.layers.dense({ units: this.depth })
    this.dense = tf.layers.dense({ units: this.depth })
    this.ffDense1 = tf.layers.dense({ units: this.depth, activation: "relu" })
    this.ffDense2 = tf.layers.dense({ units: this.depth })
    this.dropout = tf.layers.dropout({ rate: 0.2 })
    this.norm1 = tf.layers.layerNormalization()
    this.norm2 = tf.layers.layerNormalization()
  }

  computeOutputShape(inputShape: tf.Shape): tf.Shape { return [inputShape[0], this.depth] }

  call(inputs: tf.Tensor[], kwargs: { [key: string]: any }): tf.Tensor | tf.Tensor[]
  {
    let input: tf.Tensor
    if (Array.isArray(inputs)) { input = inputs[0] }
    return tf.tidy(() =>
    {
      // MultiHead Attention
      const attention = this.multiHeadAttention(inputs[0], inputs[0], inputs[0])

      const scaledInput = this.inputDense.apply(inputs[0]) as tf.Tensor
      const normalizedLatent = this.norm1.apply(tf.add(attention, scaledInput)) as tf.Tensor

      // FeedForward
      const ff1 = this.ffDense1.apply(normalizedLatent) as tf.Tensor
      const ff2 = this.ffDense2.apply(ff1)
      const dff2 = this.dropout.apply(ff2) as tf.Tensor

      // Add&Norm
      const output = this.norm2.apply(tf.add(normalizedLatent, dff2)) as tf.Tensor
      const newFeat = output.mean(1) // I know it should return the 512 tokens but I'm using it as a feature extractor here

      return newFeat
    })

  }

  getConfig(): tf.serialization.ConfigDict
  {
    const config = super.getConfig()
    Object.assign(config, {
      numHeads: this.numHeads,
      depth: this.depth
    })
    return config
  }
}

tf.serialization.registerClass(TransformerLayer)
export const transformerLayer = (args?: TransformerLayerArgs) =>
{
  return new TransformerLayer({ numHeads: args?.numHeads || 8, depth: args?.depth || 512 })
}

export default TransformerLayer;
</code>
</pre>
</details>

**Comment by gaikwadrahul8** [comment]
Hi, @ierezell 

I apologize for the delayed response. We are currently revisiting the older issues and checking if they have been resolved. just to confirm, are you still working on this feature request? If yes, I would welcome a pull request from your end. Thank you!

**Comment by ierezell** [comment]
Hello @gaikwadrahul8, 

I indeed would like this PR to be added but as I haven't got feedback from years it kinda was forgotten on my end. 

However, if you (or the team) are still willing to get attention in TFjs I can revisit / modify to conform to todays standards
